src/LiveComponent/assets/src/Component/plugins/QueryStringPlugin.ts

```diff
@@ -0,0 +1,100 @@
+import Component from '../index';
+import { PluginInterface } from './PluginInterface';
+import {
+    setQueryParam, removeQueryParam,
+} from '../../url_utils';
+
+type QueryMapping  = {
+    name: string,
+}
+
+export default class implements PluginInterface {
+    private mapping = new Map<string,QueryMapping>;
+    private initialPropsValues  = new Map<string, any>;
+    private changedProps: {[p: string]: boolean} = {};
+
+    constructor(mapping: {[p: string]: any}) {
+        Object.entries(mapping).forEach(([key, config]) => {
+            this.mapping.set(key, config);
+        })
+    }
+
+    attachToComponent(component: Component): void {
+        component.on('connect', (component: Component) => {
+            // Store initial values of mapped props
+            for (const model of this.mapping.keys()) {
+                for (const prop of this.getNormalizedPropNames(component.valueStore.get(model), model)) {
+                    this.initialPropsValues.set(prop, component.valueStore.get(prop));
+                }
+            }
+        });
+
+        component.on('render:finished', (component: Component) => {
+            this.initialPropsValues.forEach((initialValue, prop) => {
+                const value = component.valueStore.get(prop);
+
+                // Only update the URL if the prop has changed
+                this.changedProps[prop] ||= JSON.stringify(value) !== JSON.stringify(initialValue);
+                if (this.changedProps) {
+                    this.updateUrlParam(prop, value);
+                }
```

Comment:
That makes sense to me! I didn't consider the drawbacks of the current implementation for the push history strategy.

But for the last part I'm not sure. If you have a prop that is initialized on the PHP side, but not changed on client side, we should not show it in the URL (unless we have the `keep` option enable). I think we still need to track the changes in the plugin to prevent a such prop to appear in the URL after another prop changed.

For example: 

```php
#[AsLiveComponent]
class MyComponent {
    #[LiveProp(writable: true, url: true)]
    public ?string $search = null;
    
    #[LiveProp(writable: true, url: true)]
    public string $mode = 'fulltext';
}
```

If we update `search` on client side, we should get the `?search=some+string` URL. But if we don't track the initial state of `mode`, as it's not empty, we'll get `?search=some+string&mode=fulltext`. I think that's not expected, wdyt?

Reply:
Hmm. I see your point, but I'm not sure we should care. I think (?) the main thing that people will want to avoid is: "when I go to `/foo`, LiveComponents suddenly changes the URL to `/foo?search=bar&mode=fulltext` without the user doing anything". Once the user *does* change one of the `url` props, I don't think it's a big deal to have them all in the URL. For me, it's like submitting a "search" form. Even if I only type into the "search" box (but never change `mode`), I expect both `search` and `mode` to be in the query params.

Also... not tracking really simplifies things :)

Reply:
Alright, I removed tracking from the plugin :wink: